src/Microsoft.AspNet.Mvc.Core/ModelBinding/CompositeModelBinder.cs

```diff
@@ -39,6 +40,12 @@ public CompositeModelBinder([NotNull] IEnumerable<IModelBinder> modelBinders)
                 !string.IsNullOrEmpty(bindingContext.ModelName);
 
             var newBindingContext = CreateNewBindingContext(bindingContext, bindingContext.ModelName);
+            if (newBindingContext == null)
+            {
+                // Unable to find a value provider for this binding source. Binding will fail.
+                return null;
+            }
```

Comment:
This isn't an optimization. It's about the `IBindingSourceValueProvider` contract. The only possible "empty" (i.e. useless) value provider is a `CompositeValueProvider` with an empty collection. An `IBindingSourceValueProvider` implementation is supposed to return `null` instead of such a provider.

And it's fairly easy to make happen during development: Create a new non-greedy `BindingSource`, use it, but don't add an associated  factory to `MvcOptions.ValueProviderFactories`. Our tests cover this scenario, just like they cover not adding a needed `IModelBinder` to `MvcOptions.ModelBinders`.